File: OCMDAdDemo/OCMDAdDemo/DemoControllers/RewardVideoViewController.m
//
//  RewardVideoViewController.m
//  OCMDAdDemo
//
//  Created by DerekYuYi on 2019/11/14.
//  Copyright © 2019 RuiYu. All rights reserved.
//

#import "RewardVideoViewController.h"
@import MDAd;

@interface RewardVideoViewController ()<RYRewardVideoViewControllerDelegate>

@property (strong, nonatomic) RYRewardVideoViewController *rewardVideoVC;

/// 默认第一次进入页面预加载视频.
/// 如果后续需要在当前页面多次加载视频, 请在第一次预加载成功后, 设置为 NO.
@property (assign, nonatomic) BOOL isPreLoadRewardVideo;

@end

@implementation RewardVideoViewController

- (void)dealloc {
    self.rewardVideoVC.delegate = nil;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.isPreLoadRewardVideo = YES;
    
    self.rewardVideoVC = [[RYRewardVideoViewController alloc] init];
    self.rewardVideoVC.adsID = @"870001";
    self.rewardVideoVC.needDisplayBackgroundImageView = true;
    self.rewardVideoVC.delegate = self;
    [self.rewardVideoVC loadRequest];
}


- (IBAction)openRewardVideo:(UIButton *)sender {
    
    if (!self.isPreLoadRewardVideo) {
        [self.rewardVideoVC loadRequest];
    }
    
    self.rewardVideoVC.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:self.rewardVideoVC animated:true completion:nil];
    self.isPreLoadRewardVideo = NO;
    
    /*
     第二种加载方式:
     在每个场景直接创建局部 `RYRewardVideoViewController`的实例, 每个实例管理一次完整的视频加载生命周期.
     
    RYRewardVideoViewController *rewardVideoVC = [[RYRewardVideoViewController alloc] init];
    rewardVideoVC.adsID = @"870001";
    rewardVideoVC.needDisplayBackgroundImageView = true;
    rewardVideoVC.delegate = self;
    [rewardVideoVC loadRequest];
    [self presentViewController:rewardVideoVC animated:true completion:nil];
    */
}


#pragma mark - RYRewardVideoViewControllerDelegate

- (void)rewardVideoDidReceiveAd:(RYRewardVideoViewController *)controller {
    NSLog(@"reward video received successfully");
}

- (void)rewardVideoDidFailToReceiveAd:(RYRewardVideoViewController *)controller error:(RYError *)error {
    
    NSLog(@"Reward video received failed and the error info is %@", error.errorDescription);
}

- (void)rewardVideoIsReadyForPlay:(RYRewardVideoViewController *)controller isReadyForPlay:(BOOL)isReady {
    NSLog(@"rewardVideoIsReadyForPlay");
}

- (void)rewardVideoWillClose {
    [self dismissViewControllerAnimated:true completion:nil];
}

@end

File: SwiftMDAdDemo/SwiftMDAdDemo/Controllers/RewardVideoViewController.swift
//
//  RewardVideoViewController.swift
//  SwiftMDAdDemo
//
//  Created by DerekYuYi on 2019/11/5.
//  Copyright © 2019 RuiYu. All rights reserved.
//

import UIKit
import MDAd

class RewardVideoViewController: UIViewController {
    
    fileprivate let rewardVideoVC = RYRewardVideoViewController()
    
    /// 默认第一次进入页面预加载视频.
    /// 如果后续需要在当前页面多次加载视频, 请在第一次预加载成功后, 设置为 false.
    var isPreLoadRewardVideo = true
    
    deinit {
        rewardVideoVC.delegate = nil
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        rewardVideoVC.adsID = "870001"
        rewardVideoVC.needDisplayBackgroundImageView = true
        rewardVideoVC.delegate = self
        rewardVideoVC.loadRequest()
    }
    
    @IBAction func openRewardVideo(_ sender: UIButton) {
        if !isPreLoadRewardVideo {
            rewardVideoVC.loadRequest()
        }
        rewardVideoVC.modalPresentationStyle = .fullScreen
        present(rewardVideoVC, animated: true, completion: nil)
        isPreLoadRewardVideo = false
        
        /*
        第二种加载方式
        在每个场景直接创建局部 `RYRewardVideoViewController`的实例, 每个实例管理一次完整的视频加载生命周期.
        let rewardVideoVC = RYRewardVideoViewController()
        rewardVideoVC.adsID = "870001"
        rewardVideoVC.needDisplayBackgroundImageView = true
        rewardVideoVC.delegate = self
        rewardVideoVC.loadRequest()
        present(rewardVideoVC, animated: true, completion: nil)
        */
    }
    
}


extension RewardVideoViewController: RYRewardVideoViewControllerDelegate {
    
    func rewardVideoDidReceiveAd(_ controller: RYRewardVideoViewController) {
        debugPrint("reward video received successfully")
    }
    
    func rewardVideoDidFailToReceiveAd(_ controller: RYRewardVideoViewController, error: RYError) {
        debugPrint("Reward video received failed and the error info is \(error.errorDescription ?? "无详细错误信息")")
    }
    
    func rewardVideoIsReadyForPlay(_ controller: RYRewardVideoViewController, isReadyForPlay isReady: Bool) {
        if isReady {
            debugPrint("rewardVideoIsReadyForPlay")
        }
    }
    
    func rewardVideoWillClose() {
        dismiss(animated: true, completion: nil)
    }
    
}
